### ISSUE TITLE
Model inference mode

### ISSUE THREAD

**Issue by seanmor5** [elixir-nx/axon#75]
Similar API for PyTorch: https://pytorch.org/docs/1.9.0/generated/torch.inference_mode.html?highlight=inference%20mode#torch.inference_mode

Keras inference mode is a property of the model (`training=False` or `trainable=False` I can't remember).

Unfortunately, simply doing:

```elixir
{init_fn, predict_fn} = Axon.compile(model)
```

is not really good enough.  There are important differences in behaviors of certain layers in inference mode including:

- Dropout is disabled
- BatchNorm uses the EMA of mean/variance calculated over the course of training

We may want to adjust mixed precision in inference mode as well.

It'd be nice to have an API that both "freezes" the parameters into the model, drops training-only behavior in the forward pass, and maybe performs some slight optimizations to the forward expression. One option is to introduce an option to `compile` called `:mode` which specifies inference versus training mode - although I'm not sold on it yet.

One thing to note on "freezing" parameters or inlining them into the forward pass, this pattern should probably discouraged or at the very least noted as possibly harmful:

```elixir
{init_fn, predict_fn} = Axon.compile(model)

params = init_fn.()
inference_fn = &predict_fn.(params, &1)
```

While you now have a function with parameters inlined as constants, if you can't guarantee the shape of your inputs is consistent (e.g. the batch size is the same), subsequent calls to `inference_fn` with different shapes or types will load a new potentially very large executable on to your device, quickly leading to OOM. An option around this is to abstract the parameters away by placing them on the device somewhere and holding a reference to them in some global state.
